doc/move_group_interface/src/move_group_interface_tutorial.cpp

```diff
@@ -1,394 +1,648 @@
+// Both the move_group and planning_scene are accessed through their appropriate interface.
 #include <moveit/move_group_interface/move_group_interface.h>
 #include <moveit/planning_scene_interface/planning_scene_interface.h>
-
+// We also include visual_tools and several other useful message types.
+#include <moveit_visual_tools/moveit_visual_tools.h>
 #include <moveit_msgs/DisplayRobotState.h>
 #include <moveit_msgs/DisplayTrajectory.h>
-
 #include <moveit_msgs/AttachedCollisionObject.h>
 #include <moveit_msgs/CollisionObject.h>
+// END_SUB_TUTORIAL
 
-#include <moveit_visual_tools/moveit_visual_tools.h>
+// The rviz_vizual_tools namespace is shortened for code readability.
+namespace rvt = rviz_visual_tools;
 
-int main(int argc, char** argv)
+class MoveGroupInterfaceTutorial
 {
-  ros::init(argc, argv, "move_group_interface_tutorial");
-  ros::NodeHandle node_handle;
-  ros::AsyncSpinner spinner(1);
-  spinner.start();
-
-  // BEGIN_TUTORIAL
-  //
+public:
+  MoveGroupInterfaceTutorial()
+  {
+  }
+
+  void setup_visualization()
+  {
+    // Set the location where text is diplayed to above the robot
+    text_pose.translation().z() = 1.25;
+
+    // BEGIN_SUB_TUTORIAL visualization
+    // Visualization
+    // ^^^^^^^^^^^^^
+    //
+    // Within MoveItVisualTools, remote control is an introspection tool that allows users to
+    // step through a high level script via buttons and keyboard shortcuts in RViz.
+    visual_tools.loadRemoteControl();
+
+    // A variety of commands can be used to display anything from text, points or even paths.
+    visual_tools.deleteAllMarkers();
+    visual_tools.publishText(text_pose, "MoveGroupInterface Demo", rvt::WHITE, rvt::XLARGE);
+    // Batch publishing is used to reduce the number of messages being sent to RViz for large visualizations
+    visual_tools.trigger();
+    // END_SUB_TUTORIAL
+  }
+
+  void print_basic_info()
+  {
+    // BEGIN_SUB_TUTORIAL basic_info
+    // Getting Basic Information
+    // ^^^^^^^^^^^^^^^^^^^^^^^^^
+    //
+    // Move_group holds a lot of information about the robot. Accessing it is easy.
+    // For example we can print the name of the reference frame for this robot.
+    ROS_INFO_NAMED("tutorial", "Planning frame: %s", move_group.getPlanningFrame().c_str());
+
+    // We can also print the name of the end-effector link for this group.
+    ROS_INFO_NAMED("tutorial", "End effector link: %s", move_group.getEndEffectorLink().c_str());
+
+    // We even can get a list of all the groups in the robot:
+    ROS_INFO_NAMED("tutorial", "Available Planning Groups:");
+    std::copy(move_group.getJointModelGroupNames().begin(), move_group.getJointModelGroupNames().end(),
+              std::ostream_iterator<std::string>(std::cout, ", "));
+
+    // Start the Demo
+    // ^^^^^^^^^^^^^^
+    visual_tools.prompt("Press 'next' in the RvizVisualToolsGui window to start the demo");
+    // END_SUB_TUTORIAL
+  }
+
+  void plan_joint_space_goal()
```

Comment:
This should take in a joint_goal and return the plan as well as success/failure. Execution and visualization should be done in a separate class that get's re-used. Make things modular and treat this more like a high level class for interacting with move_group.